src/Runtime.hs

```diff
@@ -275,28 +270,32 @@ computeOn objDict comp =
 
 -- | Look up the arguments to a computation, apply the computation,
 -- | and set the property in the object to the result.
-applyAndSet :: (Autofloat a) => M.Map Name (Obj' a) -> ObjComp -> CompFn a -> Obj' a -> Obj' a
+applyAndSet :: (Autofloat a) => M.Map Name (Obj' a) -> ObjComp a -> CompFn a -> Obj' a -> Obj' a
 applyAndSet objDict comp function obj =
           let (objName, objProperty, fname, args) = (oName comp, oProp comp, fnName comp, fnParams comp) in
           -- Style computations rely on the order of object inputs being the same as program arguments.
-          let (constArgs, objectArgs) = partitionEithers $ map (styExprToCompExpr objDict) args in
+          let (constArgs, objectArgs) = partitionEithers $ map (resolveObjs objDict) args in
           let res = function constArgs (concat objectArgs) in
           -- TODO: for multiple objects, might not be in right order. alphabetize?
           set objProperty obj res
 
--- | Convert a style expr to an internal type (look up variable names that correspond to objects)
--- | TODO: deal with dot accesses and get/set properties (e.g. X.radius)
-styExprToCompExpr :: (Autofloat a) => M.Map Name (Obj' a) -> S.Expr -> Either (TypeIn a) ([Obj' a])
-styExprToCompExpr objs e = case e of
-                S.IntLit i     -> Left $ TInt i
-                S.FloatLit f   -> Left $ TNum $ r2f f
-                S.StringLit s  -> Left $ TStr s
-                S.Id v         -> case lookupAll v objs of
-                                  [] -> error ("id '" ++ v ++ "' /and subobjects do(es) not exist in obj dict")
-                                  xs -> Right xs
-                S.BinOp _ _ _  -> error "computations don't support operations / TODO binops"
-                S.Cons _ _     -> error "computations don't support object constructors (?)"
-                S.CompArgs _ _ -> error "computations don't support nested computations"
+-- | Look up variable names that correspond to objects and retr
+-- TODO: complete documentation
+resolveObjs :: (Autofloat a) => M.Map Name (Obj' a) -> TypeIn a
+                             -> Either (TypeIn a) [Obj' a]
+resolveObjs objs e = case e of
+    -- TODO: type synonyms logic here
+    TAllShapes v   -> case lookupAll v objs of
+                      [] -> error ("id '" ++ v ++ "' /and subobjects do(es) not exist in obj dict")
+                      xs -> Right xs
+    TShape     v   -> case M.lookup v objs of
+                      Nothing -> error ("id '" ++ v ++ "' does not exist in obj dict: " ++ show e)
```

Comment:
I prefer to put "Nothing" cases after the "Just" cases (but I recognize that's inconsistent with the list treatment above)